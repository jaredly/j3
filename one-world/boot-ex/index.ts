import { parse } from './format';
import { kwds } from './kwds';
import { Evaluator } from './types';

export const BootExampleEvaluator: Evaluator<void, void, void> = {
    kwds: kwds,
    parse(node, cursor) {
        return parse(node, cursor);
    },
    compile(top, info) {},
    evaluate(ir, irs) {},
    infer(top, infos) {},
    print(ir, irs) {
        return { code: '// printed', sourceMap: [] };
    },
};
