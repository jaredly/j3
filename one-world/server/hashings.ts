import { IHasher } from 'hash-wasm/dist/lib/WASMInterface';
import { EvaluatorPath } from '../shared/state2';

export type Module = {
    hash?: string;
    id: string;
    submodules: Record<string, { id: string; hash: string }>;
    terms: Record<string, { id: string; hash: string; idx?: number }>;
    evaluators: EvaluatorPath[];
    assets: Record<string, { id: string; hash: string }>;
    artifacts: Record<
        string,
        {
            id: string;
            hash: string;
            evaluator: EvaluatorPath;
            kind: 'evaluator' | 'ffi' | 'backend' | 'visual';
        }
    >;
    created: number;
    updated: number;
};
export type Commit = {
    hash?: string;
    sourceDoc?: string;
    root: string;
    message: string;
    parent?: string;
    author?: string;
    created: number;
};
export type Branch = { name: string; head: string };
export const hashit = (value: string, hasher: IHasher) => {
    return hasher.init().update(value).digest('hex');
};
const keyString = (k: string) =>
    k.match(/^[a-zA-Z0-9_-]+$/) ? k : JSON.stringify(k);
const norm = (value: any): string => {
    if (value == null) return '';
    if (Array.isArray(value)) return `[${value.map(norm)}]`;
    if (typeof value === 'object') {
        return `{${Object.keys(value)
            .filter((k) => value[k] !== undefined)
            .sort()
            .map((k) => `${keyString(k)}:${norm(value[k])}`)
            .join(',')}}`;
    }
    return JSON.stringify(value);
};
export const hashModule = (module: Module) =>
    norm({
        ...module,
        hash: undefined,
        created: undefined,
        updated: undefined,
    });
export const hashCommit = (commit: Commit) =>
    norm({ ...commit, hash: undefined });
