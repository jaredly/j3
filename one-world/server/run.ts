// websockets (to stream updates to toplevels that are "open" in the document I'm editing...)
// and normal dealios
// things I need to respond to:
// - namespace listing request
// - get me this toplevel
// - get me this document, hydrated with the toplevels I'll need
// -

import { Change, FullServerState, loadState, saveChanges } from './persistence';
import { Action } from '../shared/action2';
import { update } from '../shared/update2';
import { DocSelection, DocSession } from '../shared/state2';
import { ServerWebSocket } from 'bun';
import { rid } from '../shared/rid';

export type Session = {
    ws: ServerWebSocket<unknown>;
    selections: DocSession['selections'];
};

export type ClientMessage =
    | {
          type: 'presence';
          selections: DocSelection[];
      }
    | { type: 'action'; action: Action };

export type ServerMessage =
    | {
          type: 'presence';
          id: string;
          selections: DocSelection[];
      }
    | { type: 'changes'; changes: Change[] };

const baseDirectory = './.ow-data';

let state = loadState(baseDirectory);
// let ssid = 0;
const sessions: Record<string, Session> = {};

const serverUpdate = (
    state: FullServerState,
    action: Action,
    docId: string,
): FullServerState => {
    const stage = state.stages[docId];
    if (!stage)
        throw new Error(`handling action, but stage doesnt exist ${docId}`);
    const next = update(stage, action, {
        selections: {},
        toplevels: {},
    });
    return { ...state, stages: { ...state.stages, [docId]: next } };
};

Bun.serve({
    port: process.env.PORT,
    fetch(req, server) {
        console.log(req.url);
        const url = new URL(req.url);
        if (url.pathname === '/ws') {
            const ssid = url.searchParams.get('ssid');
            if (!ssid) {
                return new Response('no ssid provided', { status: 400 });
            }
            const doc = url.searchParams.get('doc');
            if (!doc) {
                return new Response('no doc provided', { status: 400 });
            }
            // upgrade the request to a WebSocket
            if (
                server.upgrade(req, {
                    data: { ssid, doc },
                    headers: {
                        'Access-Control-Allow-Origin': '*',
                        'Access-Control-Allow-Methods':
                            'GET, POST, PUT, DELETE, OPTIONS',
                    },
                })
            ) {
                return; // do not return a Response
            }
            return new Response('Upgrade failed', { status: 500 });
        }
        if (req.method === 'GET') {
            if (url.pathname === '/docs') {
                return new Response(
                    JSON.stringify(
                        Object.values(state._documents).map((doc) => ({
                            id: doc.id,
                            title: doc.title,
                        })),
                    ),
                    {
                        status: 200,
                        headers: {
                            'Content-Type': 'application/json',
                            'Access-Control-Allow-Origin': '*',
                            'Access-Control-Allow-Methods':
                                'GET, POST, PUT, DELETE, OPTIONS',
                        },
                    },
                );
            }
            if (url.pathname === '/doc') {
                const id = url.searchParams.get('id');
                if (!id) {
                    return new Response('no id provided', { status: 400 });
                }
                if (!state._documents[id]) {
                    return new Response('Doc not found', { status: 404 });
                }
                if (!state.stages[id]) {
                    state.stages[id] = {
                        ...state._documents[id],
                        history: [],
                        toplevels: {},
                    };
                    // TODO: do we prune here? maybe?
                    Object.values(state.stages[id].nodes).forEach((node) => {
                        state.stages[id].toplevels[node.toplevel] =
                            state._toplevels[node.toplevel];
                    });
                }

                return new Response(JSON.stringify(state.stages[id]), {
                    status: 200,
                    headers: {
                        'Content-Type': 'application/json',
                        'Access-Control-Allow-Origin': '*',
                        'Access-Control-Allow-Methods':
                            'GET, POST, PUT, DELETE, OPTIONS',
                    },
                });
            }
            return new Response(JSON.stringify(state), {
                status: 200,
                headers: {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Methods':
                        'GET, POST, PUT, DELETE, OPTIONS',
                },
            });
        }
        return new Response('ok', {
            headers: {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods':
                    'GET, POST, PUT, DELETE, OPTIONS',
            },
        });
    },
    websocket: {
        open(ws) {
            let ssid: string = (ws.data as any).ssid;
            if (sessions[ssid]) {
                console.log('need new ssid', ssid, 'is taken');
                ssid = rid() + rid();
                (ws.data as any).ssid = ssid;
            }
            console.log('ws open', ssid);
            sessions[ssid] = { selections: [], ws };
            ws.send(JSON.stringify({ type: 'hello', ssid }));
        },
        close(ws) {
            const ssid: string = (ws.data as any).ssid;
            console.log('ws close', ssid);
            delete sessions[ssid];
        },
        // handler called when a message is received
        async message(ws, message) {
            const msg: ClientMessage = JSON.parse(String(message));
            const { ssid, doc } = ws.data as { ssid: string; doc: string };
            if (msg.type === 'presence') {
                sessions[ssid].selections = msg.selections;

                Object.keys(sessions).forEach((k) => {
                    if (k !== ssid) {
                        sessions[k].ws.send(
                            JSON.stringify({
                                type: 'presence',
                                selections: msg.selections,
                                id: ssid,
                            } satisfies ServerMessage),
                        );
                    }
                });
            } else if (msg.type === 'action') {
                // console.log('action', msg.action);
                // NOTE: this can be sooo much more efficient, by having update
                // also report on what changed.
                const next = serverUpdate(state, msg.action, doc);
                const changes = saveChanges(baseDirectory, state, next);

                state = next;
                // notify others of the changes
                Object.keys(sessions).forEach((k) => {
                    if (k !== ssid) {
                        sessions[k].ws.send(
                            JSON.stringify({
                                type: 'changes',
                                changes,
                            } satisfies ServerMessage),
                        );
                    }
                });
            }
        },
    },
});

console.log(`Listening on http://localhost:${process.env.PORT}`);
