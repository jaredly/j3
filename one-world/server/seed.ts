import { BunSQLiteDatabase } from 'drizzle-orm/bun-sqlite';
import { createBLAKE3 } from 'hash-wasm';
import {
    Branch,
    Commit,
    hashCommit,
    hashit,
    hashModule,
    Module,
} from './hashings';
import * as tb from './schema';

export async function seed(db: BunSQLiteDatabase<typeof tb>) {
    const hasher = await createBLAKE3();

    const module: Module = {
        id: 'root',
        aliases: {},
        submodules: {},
        terms: {},
        evaluators: [],
        assets: {},
        artifacts: {},
        created: Date.now(),
        updated: Date.now(),
    };
    const moduleHash = hashit(hashModule(module), hasher);

    const commit: Commit = {
        root: moduleHash,
        message: 'Initial',
        author: 'me',
        created: Date.now(),
    };
    const commitHash = hashit(hashCommit(commit), hasher);

    const branch: Branch = {
        name: 'main',
        head: commitHash,
    };

    await db.insert(tb.modules).values({
        ...module,
        hash: moduleHash,
        submodules: module.submodules,
        terms: module.terms,
        // documents: (module.documents),
        evaluators: module.evaluators,
    });

    await db.insert(tb.commits).values({
        ...commit,
        hash: commitHash,
    });

    await db.insert(tb.branches).values(branch);

    // await db.insert(tb.modules).values();

    //     const user: typeof usersTable.$inferInsert = {
    //         name: 'John',
    //         age: 30,
    //         email: 'john@example.com',
    //     };

    //     await db.insert(usersTable).values(user);
    //     console.log('New user created!');

    //     const users = await db.select().from(usersTable);
    //     console.log('Getting all users from the database: ', users);
    //     /*
    //   const users: {
    //     id: number;
    //     name: string;
    //     age: number;
    //     email: string;
    //   }[]
    //   */

    //     await db
    //         .update(usersTable)
    //         .set({
    //             age: 31,
    //         })
    //         .where(eq(usersTable.email, user.email));
    //     console.log('User info updated!');

    //     await db.delete(usersTable).where(eq(usersTable.email, user.email));
    //     console.log('User deleted!');
}
