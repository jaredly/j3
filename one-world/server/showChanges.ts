import { recNodeToText } from '../client/cli/drawDocNode';
import { aBlockToString } from '../shared/IR/block-to-attributed-text';
import { fromMap, Loc } from '../shared/nodes';
import { HistoryItem, DocumentNode } from '../shared/state2';
import { Toplevel } from '../shared/toplevels';

export const showChanges = (changes: HistoryItem['changes']) => {
    const text: string[] = [];

    const nt = (node: DocumentNode) =>
        `${node.id}:${node.toplevel}:${node.children.join(',')}`;
    const tt = (top: Toplevel) => {
        const rn = fromMap((n) => [[top.id, n]] as Loc, top.root, top.nodes, {
            children: [],
            map: {},
        });
        return (
            `top: ${top.id} = ` +
            aBlockToString(
                recNodeToText(
                    rn,
                    {
                        errors: [],
                        exports: [],
                        layouts: [],
                        references: [],
                        styles: {},
                        tableHeaders: {},
                        top: null,
                    },
                    100,
                ),
                false,
            )
        );
    };

    if (changes.nodes || changes.prevNodes) {
        if (changes.nodes) {
            Object.keys(changes.nodes).forEach((k) => {
                if (changes.prevNodes && changes.prevNodes[+k]) {
                    text.push(`- ${nt(changes.prevNodes[+k])}`);
                }
                text.push(`+ ${nt(changes.nodes![+k])}`);
            });
        }
        if (changes.prevNodes) {
            Object.keys(changes.prevNodes).forEach((k) => {
                if (!changes.nodes || !changes.nodes[+k]) {
                    text.push(`- ${nt(changes.prevNodes![+k])}`);
                }
            });
        }
    }

    if (changes.toplevels || changes.prevToplevels) {
        if (changes.toplevels) {
            Object.keys(changes.toplevels).forEach((k) => {
                if (changes.prevToplevels && changes.prevToplevels[k]) {
                    text.push(`- ${tt(changes.prevToplevels[k])}`);
                }
                if (changes.toplevels![k]) {
                    text.push(`+ ${tt(changes.toplevels![k])}`);
                }
            });
        }
        if (changes.prevToplevels) {
            Object.keys(changes.prevToplevels).forEach((k) => {
                if (!changes.toplevels || !changes.toplevels[k]) {
                    text.push(`- ${tt(changes.prevToplevels![k])}`);
                }
            });
        }
    }
    return text.join('\n');
};
