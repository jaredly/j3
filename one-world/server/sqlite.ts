import sqlite from 'better-sqlite3';
import { ServerBackend } from './json-git';

const ts = `created DATETIME DEFAULT CURRENT_TIMESTAMP,
    updated DATETIME DEFAULT CURRENT_TIMESTAMP,`;

const tables: string[] = [];

const top_shared = `
    module TEXT NOT NULL,

    body TEXT NOT NULL, -- jsonified nodes, root, nextLoc, auxiliaries, testConfig(??)

    ${ts}
`;

tables.push(`
CREATE TABLE IF NOT EXISTS toplevels (
    id TEXT NOT NULL,
    hash TEXT NOT NULL,

    ${top_shared}

    PRIMARY KEY (hash, id)
)
`);

tables.push(`
CREATE TABLE IF NOT EXISTS edited_documents_toplevels (
    topid TEXT NOT NULL,
    docid TEXT NOT NULL,
    hash TEXT, -- might be null if this is a new toplevel

    ${top_shared}

    PRIMARY KEY (docid, topid)
)
`);

const doc_shared = `
    title TEXT NOT NULL,
    published DATETIME,
    module TEXT NOT NULL,
    evaluator TEXT NOT NULL,

    body TEXT NOT NULL, -- jsonified nodes, nextLoc, nsAliases

    ${ts}
`;

tables.push(`
CREATE TABLE IF NOT EXISTS documents (
    id TEXT PRIMARY KEY,
    hash TEXT NOT NULL,

    ${doc_shared}

    PRIMARY KEY (hash, id)
)
`);

tables.push(`
CREATE TABLE IF NOT EXISTS edited_documents (
    id TEXT PRIMARY KEY,

    ${doc_shared}
)
`);

tables.push(`
CREATE TABLE IF NOT EXISTS edited_documents_history (
    id INTEGER PRIMARY KEY,

    ${doc_shared}
)
`);

// export const sqliteBackend = (dbPath: string): ServerBackend => {
//     const db = sqlite(dbPath);

//     db.prepare(
//         `
//       `,
//     ).run();

//     // return {};
// };
