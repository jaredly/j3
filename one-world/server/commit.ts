import 'dotenv/config';
import { and, eq, sql } from 'drizzle-orm';
import { BunSQLiteDatabase } from 'drizzle-orm/bun-sqlite';
import { genId } from '../client/cli/edit/newDocument';
import {
    DocStage,
    HashedMod,
    HistoryItem,
    LockedNode,
    Mod,
} from '../shared/state2';
import { Toplevel } from '../shared/toplevels';
import * as tb from './schema';
import { Updated } from '../shared/update2';
import { hashit, norm, normDoc } from './hashings';
import { createBLAKE3 } from 'hash-wasm';
import { DrizzleDb, getEditedDoc, getHeadRoot } from './drizzle';
import { hashToplevels } from '../graphh/just-organize';

export const updateRoot = async (
    db: DrizzleDb,
    rootHash: string,
    doc: DocStage,
) => {
    const root = await db.query.modules.findFirst({
        where: and(eq(tb.modules.id, 'root'), eq(tb.modules.hash, rootHash)),
    });
    if (!root) throw new Error(`no root`);
    if (!doc.modules[doc.id]) throw new Error(`doc doesnt have its own module`);
    const modules: Record<string, Mod> = { ...doc.modules };

    // Expanding all relevant parents
    for (let key of Object.keys(doc.modules)) {
        let pmod = null as null | Mod;
        const path: string[] = [];
        for (let parent of doc.modules[key].path) {
            if (!modules[parent]) {
                if (!pmod) throw new Error(`cant get tto ${parent}`);
                let self = Object.values(pmod.submodules).find(
                    (m) => m.id === parent,
                );
                if (!self)
                    throw new Error(
                        `self ${parent} not found in parent ${pmod.id}`,
                    );
                const got = await db.query.modules.findFirst({
                    where: and(
                        eq(tb.modules.id, parent),
                        eq(tb.modules.hash, self.hash),
                    ),
                });
                if (!got)
                    throw new Error(
                        `module not in db ${parent} with hash ${self.hash}`,
                    );
                modules[parent] = {
                    ...got,
                    path: path.slice(),
                    ts: { created: got.created, updated: got.updated },
                };
            } else {
                pmod = modules[parent];
            }
            path.push(parent);
        }
    }

    const hasher = await createBLAKE3();
    const hashed: Record<string, HashedMod> = {};
    const process = (id: string): string => {
        let mod = modules[id];
        if (!mod) throw new Error('no mod');
        mod = { ...mod, submodules: { ...mod.submodules } };
        Object.keys(mod.submodules).forEach((key) => {
            const { id: child } = mod.submodules[key];
            if (modules[child]) {
                mod.submodules[key] = { id: child, hash: process(child) };
            }
        });

        let docHash = mod.docHash;
        if (id === doc.id) {
            if (!doc.hash) throw new Error(`doc needs to be hashed first`);
            docHash = doc.hash;
        }
        if (docHash == null) {
            throw new Error(`mod doesnt have a doc hash`);
        }

        const hash = hashit(normMod(mod), hasher);
        hashed[id] = { ...mod, hash, docHash };
        return hash;
    };
    process('root');

    await db.insert(tb.modules).values(
        Object.values(hashed).map((mod) => ({
            ...mod,
            created: mod.ts.created,
            updated: mod.ts.updated,
        })),
    );

    return hashed.root.hash;
};

const normMod = (mod: Mod) => {
    return norm({ ...mod, hash: undefined, ts: undefined });
};

const normModule = (mod: typeof tb.modules.$inferSelect) => {
    return norm({
        ...mod,
        hash: undefined,
        created: undefined,
        updated: undefined,
    });
};

export const commitDoc = async (
    db: DrizzleDb,
    id: string,
    branch: string,
    hash: (raw: string) => string,
) => {
    let ds = await getEditedDoc(db, id, branch);
    const root = await getHeadRoot(db, branch);
    if (!ds) throw new Error(`doc not edited ${id}`);
    if (root !== ds.root) throw new Error(`need to merge before committing!`);

    // Hash toplevels
    const tops = hashToplevels(ds.toplevels, (tops) =>
        hash(
            norm(
                tops.map((t) => ({ ...t, hash: undefined, module: undefined })),
            ),
        ),
    );

    // Commit toplevels
    await db.insert(tb.toplevels).values(
        Object.values(tops).map(({ nodes, nextLoc, root, ts, ...top }) => ({
            ...top,
            updated: ts.updated,
            created: ts.created,
            hash: top.hash!,
            body: { nodes, nextLoc, root },
            // TODO fill this out
            accessories: [],
        })),
    );

    // Hash doc
    const nodes: Record<string, LockedNode> = {};
    Object.values(ds.nodes).forEach((node) => {
        nodes[node.id] = {
            ...node,
            topLock: {
                hash: tops[node.toplevel].hash,
                manual: false,
            },
        };
    });
    const docHash = hash(
        norm({
            id: ds.id,
            title: ds.title,
            published: ds.published,
            evaluator: ds.evaluator,
            nodes,
            nextLoc: ds.nextLoc,
            ts: undefined,
        }),
    );

    // Commit doc
    await db
        .insert(tb.documents)
        .values([
            {
                id: ds.id,
                hash: docHash,
                title: ds.title,
                published: ds.published,
                evaluator: ds.evaluator,
                body: {
                    nodes,
                    nextLoc: ds.nextLoc,
                },
                updated: ds.ts.updated,
                created: ds.ts.created,
                archived: ds.archived,
            },
        ])
        .onConflictDoNothing();

    // Commit modules
    ds = { ...ds, toplevels: tops, hash: docHash };
    ds.modules = { ...ds.modules };
    ds.modules[ds.id] = { ...ds.modules[ds.id], docHash };
    const nroot = await updateRoot(db, root, ds);
    return nroot;
};

// hrm
