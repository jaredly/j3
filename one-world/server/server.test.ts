import sqlite from 'bun:sqlite';
import { eq } from 'drizzle-orm';
import { drizzle } from 'drizzle-orm/bun-sqlite';

import { expect, test } from 'bun:test';
import * as dk from 'drizzle-kit/payload';
import { splitGraphemes } from '../../src/parse/splitGraphemes';
import {
    handleDropdown,
    handleMovement,
    handleUpDown,
} from '../client/cli/edit/handleMovement';
import { handleUpdate } from '../client/cli/edit/handleUpdate';
import { parseAndCache, render, RState } from '../client/cli/render';
import { newStore, WS } from '../client/newStore2';
import { AnyEvaluator } from '../evaluators/boot-ex/types';
import { NopEvaluator, SimplestEvaluator } from '../evaluators/simplest';
import { Caches, Context, evaluate } from '../graphh/by-hand';
import { aBlockToString } from '../shared/IR/block-to-attributed-text';
import { ParseAndEval, selectEODoc } from '../shared/IR/nav';
import { DocStage } from '../shared/state2';
import {
    DrizzleDb,
    getDoc,
    getEditedDoc,
    getHeadRoot,
    getModule,
    newStage,
    saveDocument,
} from './drizzle';
import * as tb from './schema';
import { seed } from './seed';
import { showChanges } from './showChanges';
import { ServerMessage } from './run';
import { applyChanges } from '../shared/update2';
import { commitDoc } from './commit';
import { createBLAKE3 } from 'hash-wasm';
import { hashit } from './hashings';

const prepare = (async () => {
    const prev = await dk.generateSQLiteDrizzleJson({});
    const schema = await dk.generateSQLiteDrizzleJson(tb);
    return dk.generateSQLiteMigration(prev, schema);
})();

const emptyDb = async (name = ':memory:'): Promise<DrizzleDb> => {
    const db = sqlite.open(name);

    db.exec('begin');
    (await prepare).forEach((stmt) => db.exec(stmt));
    db.exec('commit');

    const dzb = drizzle(db, { schema: tb });
    await seed(dzb);
    return dzb;
};

test('doing a little seed I guess', async () => {
    const db = await emptyDb();

    const main = await db.query.branches.findFirst({
        where: eq(tb.branches.name, 'main'),
        with: { commit: { with: { module: true } } },
    });
    expect(main).toBeTruthy();
    const { commit } = main!;
    expect(commit.module.id).toEqual('root');
});

/*

Full end-to-end:

- clean seed
- new document (at root & main)
- edit the root toplevel with some text
- commit it
- load up and serialize the document in its new form

*/

const nows: WS = { close() {}, onMessage(fn) {}, send(msg) {} };

const reval = (
    caches: Caches<unknown>,
    ctx: Context,
    rstate: RState,
    ev: AnyEvaluator,
) => {
    Object.keys(caches.parse).forEach((tid) => {
        rstate.parseAndEval[tid].output = evaluate(tid, ctx, ev, caches);
    });
};

const specials: { [key: string]: string } = {
    ArrowUp: 'UP',
    ArrowDown: 'DOWN',
    ArrowLeft: 'LEFT',
    ArrowRight: 'RIGHT',
    Enter: 'ENTER',
    Backspace: 'BACKSPACE',
    Tab: 'TAB',
    Undo: 'UNDO',
    Redo: 'REDO',
};

const runText = (
    ds: DocStage,
    text: string | string[],
    ws: WS = nows,
    ev: AnyEvaluator = NopEvaluator,
) => {
    const store = newStore(ds!, ws, null);

    const W = 200;

    const { parseCache, caches, ctx } = parseAndCache(store, ds.id, {}, ev);
    updateModuleExports(parseCache);
    let rstate = render(W, store, ds.id, parseCache);

    // Select the end of the document!
    const sel = selectEODoc(ds!, rstate.cache);
    expect(sel).toBeTruthy();
    store.update({ type: 'selection', doc: ds.id, selections: [sel!] });

    reval(caches, ctx, rstate, ev);

    const singles: string[] = [];
    if (Array.isArray(text)) {
        text.slice(1).forEach((line) => {
            if (specials[line]) {
                singles.push(specials[line]);
            } else {
                singles.push(...splitGraphemes(line));
            }
        });
    } else {
        singles.push(...splitGraphemes(text));
    }

    function updateModuleExports(parseCache: ParseAndEval<unknown>) {
        const state = store.getState();
        Object.entries(parseCache).forEach(([tid, value]) => {
            const top = state.toplevels[tid];
            value.parseResult.exports.forEach((exp) => {
                const name = top.nodes[exp.loc[0][1]];
                if (name.type !== 'id') {
                    throw new Error(`export loc must be id`);
                }
                if (name.ref) {
                    throw new Error(`export id must not have a ref`);
                }
                state.modules[top.module].terms[name.text] = {
                    hash: '',
                    id: tid,
                    idx: exp.loc[0][1],
                };
            });
        });
        // TODO remove staleee
    }

    singles.forEach((key) => {
        const { parseCache, caches, ctx } = parseAndCache(store, ds.id, {}, ev);
        // state.modules ... are going to be mutable for now.
        updateModuleExports(parseCache);
        rstate = render(W, store, ds.id, parseCache);

        if (handleDropdown(key, ds.id, store, rstate, () => {}, ev)) {
            return true;
        }

        if (
            handleUpDown(key, ds.id, store, rstate) ||
            handleMovement(key, ds.id, rstate.cache, store)
        ) {
            return true;
        }

        if (handleUpdate(key, ds.id, rstate.cache, store)) {
            return true;
        }
    });

    store.update({ type: 'selection', doc: ds.id, selections: [] });

    {
        const { parseCache, caches, ctx } = parseAndCache(store, ds.id, {}, ev);
        // state.modules ... are going to be mutable for now.
        updateModuleExports(parseCache);
    }

    return store.getState();
};

const editorToString = (ds: DocStage, W = 200) => {
    const store = newStore(ds!, nows, null);
    const { parseCache, caches, ctx } = parseAndCache(
        store,
        ds.id,
        {},
        NopEvaluator,
    );
    let rstate = render(W, store, ds.id, parseCache, {
        plainBullets: true,
        showRefHashes: true,
    });
    reval(caches, ctx, rstate, NopEvaluator);
    rstate = render(W, store, ds.id, parseCache, {
        plainBullets: true,
        showRefHashes: true,
    });
    return aBlockToString(rstate.txt, false);
};

/*

Let's talk, tests.


DS -> keys -> DS


*/

// MARK: New tests

test.only("nowww let's do some exportsss", async () => {
    const text = 'x 1 (def x 10) x 23';
    const doc = newStage('lol', 10, 'na');
    const d2 = runText(doc, text, undefined, SimplestEvaluator);
    // ok forks. how do we populate the modules.
    // expect(editorToString(d2, 200)).toEqual(getEx(text));
    expect(d2.modules[d2.id].terms.x).toBeTruthy();
});

// DS -> DB -> DS
test('store and retrieve a newStage', async () => {
    const db = await emptyDb();
    const root = await getHeadRoot(db, 'main');
    const doc = newStage('lol', 10, root);
    await saveDocument(db, doc, 'main');
    const got = await getEditedDoc(db, doc.id, 'main');
    expect(got).toEqual(doc);
});

const getEx = (text: string | string[]) =>
    Array.isArray(text) ? text[0] : '- ' + text;

const examples: (string | string[])[] = [
    'hello',
    '[one two]',
    '{yes no 123}',
    '(hello [folks yes])',
    [
        '- (hello forks)',
        '(hello folk',
        'ArrowLeft',
        'Backspace',
        'r',
        'ArrowRight',
        's)',
    ],
    // ['- hello-yall\n- folks', 'hello folks', 'ArrowUp', '-yall'],
];

test('lets do some history', () => {
    const text = ['- (hello)', '(hello folks', 'Undo'];
    // const text = ['(hello)', '(hello last things', 'Undo'];
    const doc = newStage('lol', 10, 'na');
    const d2 = runText(doc, text);
    expect(
        d2.history.map((h) => showChanges(h.changes)).join('\n\n'),
    ).toMatchSnapshot();
    expect(editorToString(d2, 200)).toEqual(getEx(text));
});

test('lets do some movement and stuff', () => {
    const text = ['- hello-yall\n- folks', 'hello folks', 'ArrowUp', '-yall'];
    const doc = newStage('lol', 10, 'na');
    const d2 = runText(doc, text);
    expect(
        d2.history.map((h) => showChanges(h.changes)).join('\n\n'),
    ).toMatchSnapshot();
    expect(editorToString(d2, 200)).toEqual(getEx(text));
});

// DS -> keys -> DS
examples.forEach((text) => {
    test('DS -> keys -> DS : ' + getEx(text), () => {
        const doc = newStage('lol', 10, 'na');
        const d2 = runText(doc, text);
        expect(editorToString(d2, 200)).toEqual(getEx(text));
    });
});

// DS -> keys&cross
// TODO: revisit once I have history items

examples.forEach((text) => {
    test('DS sync to other DS : ' + getEx(text), async () => {
        const doc = newStage('lol', 10, 'na');
        const listeners: ((msg: ServerMessage) => void)[] = [];

        const other = newStore(
            doc,
            {
                close() {},
                onMessage(fn) {
                    listeners.push(fn);
                },
                send(msg) {},
            },
            null,
        );

        const d2 = runText(doc, text, {
            close() {},
            onMessage(fn) {},
            send(msg) {
                if (msg.type === 'changes') {
                    listeners.forEach((f) => f(msg));
                }
            },
        });
        expect(editorToString(d2, 200)).toEqual(getEx(text));
        expect(editorToString(other.getState(), 200)).toEqual(getEx(text));
    });
});

examples.forEach((text) => {
    test('DS->DB->DS -> keys -> DS->DB->DS : ' + getEx(text), async () => {
        const db = await emptyDb();
        const root = await getHeadRoot(db, 'main');

        const id = 'lol';
        await saveDocument(db, newStage(id, 10, root), 'main');

        const doc = runText((await getEditedDoc(db, id, 'main'))!, text);

        await saveDocument(db, doc, 'main');

        expect(
            editorToString((await getEditedDoc(db, id, 'main'))!, 200),
        ).toEqual(getEx(text));
    });
});

const queue = () => {
    let cur = Promise.resolve<unknown>(0);
    return {
        enqueue(p: () => Promise<unknown>) {
            cur = cur.then(p);
        },
        fin() {
            return cur;
        },
    };
};

examples.forEach((text) => {
    test('DS -> keys & sync : ' + getEx(text), async () => {
        const id = 'lol';
        const doc = newStage(id, 10, 'main');
        let tmp = doc;

        const edit = runText(doc, text, {
            close() {},
            onMessage(fn) {},
            send(msg) {
                if (msg.type === 'changes') {
                    // const ups: Updated = { toplevels: {}, selections: {} };
                    msg.items.forEach((item) => {
                        tmp = applyChanges(tmp, item.changes);
                    });
                    tmp = { ...tmp, history: tmp.history.concat(msg.items) };
                }
            },
        });

        expect(edit).toEqual(tmp);
        expect(editorToString(tmp, 200)).toEqual(getEx(text));
    });
});

examples.forEach((text) => {
    test('DS->DB->DS -> keys & sync -> DB->DS : ' + getEx(text), async () => {
        const db = await emptyDb();
        const root = await getHeadRoot(db, 'main');
        const id = 'lol';

        await saveDocument(db, newStage(id, 10, root), 'main');
        const doc = (await getEditedDoc(db, id, 'main'))!;

        const { enqueue, fin } = queue();

        const edit = runText(doc, text, {
            close() {},
            onMessage(fn) {},
            send(msg) {
                if (msg.type === 'changes') {
                    enqueue(async () => {
                        const doc = await getEditedDoc(db, id, 'main');
                        let upd = doc!;
                        if (msg.items[0].idx !== upd.history.length) {
                            throw new Error(`history out of sync`);
                        }
                        msg.items.forEach((item) => {
                            upd = applyChanges(upd, item.changes);
                        });
                        upd = {
                            ...upd,
                            history: upd.history.concat(msg.items),
                        };
                        await saveDocument(db, upd, 'main');
                    });
                }
            },
        });

        await fin();

        const dbdoc = await getEditedDoc(db, id, 'main');
        expect(dbdoc).toEqual(edit);
        expect(editorToString(dbdoc!, 200)).toEqual(getEx(text));
    });
});

test('now try to commit', async () => {
    const db = await emptyDb();
    const root = await getHeadRoot(db, 'main');

    const id = 'lol';
    await saveDocument(db, newStage(id, 10, root), 'main');

    const text = 'hello multiple folks';
    const doc = runText((await getEditedDoc(db, id, 'main'))!, text);

    await saveDocument(db, doc, 'main');

    const hasher = await createBLAKE3();
    const nroot = await commitDoc(db, id, 'main', (str) => hashit(str, hasher));

    const docModule = await getModule(db, nroot, [doc.id]);
    const docBack = await getDoc(db, doc.id, docModule!.docHash, nroot);

    expect(editorToString(docBack!, 200)).toEqual(
        '- hello\n- multiple\n- folks',
    );
});

//
// old, silly tests

// test('noww to like, make a new document?', async () => {
//     const db = await emptyDb();
//     const root = await getHeadRoot(db, 'main');
//     const id = await newDocument(db, root, 'main');
//     const ds = await getEditedDoc(db, id, 'main');
//     expect(ds).toBeTruthy();

//     const text = editorToString(runText(ds!, 'hello-folks'));

//     expect(text).toEqual('- hello-folks');
// });

// const fullRun = async (text: string) => {
//     const db = await emptyDb();
//     const root = await getHeadRoot(db, 'main');
//     const id = await newDocument(db, root, 'main');
//     const ds = await getEditedDoc(db, id, 'main');
//     expect(ds).toBeTruthy();

//     return editorToString(runText(ds!, text));
// };

/*

Fuller ender-to-end

- clean seed
- new document
- edit with `(def x 10)`
- commit it
- /x should be defined

*/
