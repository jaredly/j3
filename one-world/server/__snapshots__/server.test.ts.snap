// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`lets do some history 1`] = `
"- top: lol:top = 
+ top: lol:top = ()

- top: lol:top = ()
+ top: lol:top = (hello)

- top: lol:top = (hello)
+ top: lol:top = (hello )

- top: lol:top = (hello )
+ top: lol:top = (hello)"
`;

exports[`lets do some movement and stuff 1`] = `
"- top: lol:top = 
+ top: lol:top = hello
+ top: rko36eaqyh = 

- top: rko36eaqyh = 
+ top: rko36eaqyh = folks

- top: lol:top = hello
+ top: lol:top = hello-yall"
`;
